sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotShiftByZeroOrIntSize.cs

```diff
@@ -1,151 +1,238 @@
+            var binaryExpression = node as BinaryExpressionSyntax;
+            if (binaryExpression?.OperatorToken.IsKind(SyntaxKind.LessThanLessThanToken) ?? false)
+            {
+                return new Tuple<bool, ExpressionSyntax>(true, binaryExpression.Right);
+            }
+
+            if (binaryExpression?.OperatorToken.IsKind(SyntaxKind.GreaterThanGreaterThanToken) ?? false)
+            {
+                return new Tuple<bool, ExpressionSyntax>(false, binaryExpression.Right);
+            }
+
+            var assignmentExpession = node as AssignmentExpressionSyntax;
+            if (assignmentExpession?.OperatorToken.IsKind(SyntaxKind.LessThanLessThanEqualsToken) ?? false)
+            {
+                return new Tuple<bool, ExpressionSyntax>(true, assignmentExpession.Right);
+            }
+
+            if (assignmentExpession?.OperatorToken.IsKind(SyntaxKind.GreaterThanGreaterThanEqualsToken) ?? false)
+            {
+                return new Tuple<bool, ExpressionSyntax>(false, assignmentExpession.Right);
+            }
+
+            return null;
+        }
+
+        private static bool TryGetConstantValue(ExpressionSyntax expression, out int value)
+        {
+            value = 0;
+            var literalExpression = expression?.RemoveParentheses() as LiteralExpressionSyntax;
+            return literalExpression != null &&
+                int.TryParse(literalExpression.Token.ValueText, out value);
+        }
+
+        private static ShiftInstance FindIssues(SyntaxNode node, SemanticModel semanticModel)
+        {
+            var tuple = GetRhsArgumentOfShiftNode(node);
+            if (tuple == null)
+            {
+                return null;
+            }
+
+            if (!TryGetConstantValue(tuple.Item2, out var shiftByCount))
+            {
+                return new ShiftInstance(node);
+            }
+
+            var typeSymbol = semanticModel.GetTypeInfo(node).ConvertedType;
+            if (typeSymbol == null)
+            {
+                return new ShiftInstance(node);
+            }
+
+            var variableBitLength = FindTypeSizeOrDefault(typeSymbol);
+            if (variableBitLength == 0)
+            {
+                return new ShiftInstance(node);
+            }
+
+            var issueDescription = FindProblemDescription(variableBitLength, shiftByCount, tuple.Item1, out bool isLiteralZero);
+            return issueDescription == null ? new ShiftInstance(node) : new ShiftInstance(issueDescription, isLiteralZero, node);
+        }
+
+        private static int FindTypeSizeOrDefault(ITypeSymbol typeSymbol)
+        {
+            return mapKnownTypesToIntegerBitSize
+                .Where(kv => typeSymbol.Is(kv.Key))
+                .Select(kv => kv.Value)
+                .FirstOrDefault();
+        }
+
+        private static string FindProblemDescription(int typeSizeInBits, int shiftBy, bool isLeftShift, out bool isLiteralZero)
+        {
+            isLiteralZero = false;

```

Comment:
I would move this line after the first `if` as you initialize it to `false` within the `if`.